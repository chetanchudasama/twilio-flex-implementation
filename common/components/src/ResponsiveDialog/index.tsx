import Button from "@material-ui/core/Button";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogTitle from "@material-ui/core/DialogTitle";
import React from "react";
import styled from "react-emotion";
import { IconButton, Icon } from "@material-ui/core";
import CloseIcon from "@material-ui/icons/Close";
import { ResponsiveDialogStyles } from "./ResponsiveDialog.Styles";
import { white, secondary, deleteButtonColor } from "../shared/AppTheme";

export interface ResponsiveDialogProps {
  open: boolean;
  title?: string;
  okText?: string;
  cancelText?: string;
  deleteText?: string;
  maxWidth?: "xs" | "sm" | "md" | "lg" | "xl" | false;
  showActionButtons?: boolean;
  onConfirm: () => void;
  onCancel: () => void;
  isConfirmationDialog?: boolean;
  disableSaveButton?: boolean;
  titleCloseIcon?: boolean;
  showDeleteButton?: boolean;
  onDelete?: () => void;
}

export const ResponsiveDialog: React.FC<ResponsiveDialogProps> = ({
  maxWidth,
  open,
  onCancel,
  onConfirm,
  title,
  titleCloseIcon,
  isConfirmationDialog,
  showActionButtons,
  cancelText,
  okText,
  deleteText,
  children,
  showDeleteButton,
  disableSaveButton,
  onDelete,
}) => {
  const StyledSaveButton = styled(Button)({
    backgroundColor: `${secondary}!important`,
    color: `${white}!important`,
  });

  const StyledCancelButton = styled(Button)({
    backgroundColor: `${white}!important`,
    color: `${secondary}!important`,
  });

  const StyledDeleteButton = styled(Button)({
    backgroundColor: `${white}!important`,
    color: `${deleteButtonColor}!important`,
  });

  return (
    <ResponsiveDialogStyles
      fullWidth
      maxWidth={maxWidth || "sm"}
      open={open}
      onClose={onCancel}
      aria-labelledby="max-width-dialog-title"
      disableBackdropClick
      disableEscapeKeyDown
      scroll="paper"
      disableEnforceFocus
    >
      {title || titleCloseIcon ? (
        <DialogTitle
          className={!titleCloseIcon ? "dialog-title" : ""}
          id="max-width-dialog-title"
        >
          {title || ""}
          {titleCloseIcon ? (
            <IconButton
              aria-label="Close"
              className="close-icon"
              onClick={onCancel}
            >
              <CloseIcon />
            </IconButton>
          ) : null}
        </DialogTitle>
      ) : null}
      <DialogContent
        className={isConfirmationDialog ? "confirmation-message" : ""}
      >
        {children}
      </DialogContent>
      {showActionButtons && (
        <DialogActions>
          {showDeleteButton && (
            <StyledDeleteButton
              variant="text"
              onClick={onDelete}
              type="button"
              className="delete-btn"
            >
              <Icon className="delete-btn-icon">delete</Icon>
              {deleteText || "Delete"}
            </StyledDeleteButton>
          )}
          <StyledCancelButton variant="text" onClick={onCancel} type="button">
            {cancelText || "Close"}
          </StyledCancelButton>
          <StyledSaveButton
            variant="contained"
            type="button"
            onClick={onConfirm}
            className={disableSaveButton ? "disabled-btn" : ""}
          >
            {okText || "Save"}
          </StyledSaveButton>
        </DialogActions>
      )}
    </ResponsiveDialogStyles>
  );
};
