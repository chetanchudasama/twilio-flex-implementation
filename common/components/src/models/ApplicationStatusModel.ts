export class ApplicationStatusModel {
  applicationStatusId: number;

  applicationStatusName: string;

  constructor() {
    this.applicationStatusId = 0;
    this.applicationStatusName = "";
  }
}
