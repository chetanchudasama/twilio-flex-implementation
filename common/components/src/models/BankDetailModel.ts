export class BankDetailModel {
  public bankName: string;

  public monthsAtBank: number;

  public nameOnBankAccount: string;

  public accountNumber: string;

  public sortCode: string;

  public netMonthlyIncome: number;

  constructor() {
    this.bankName = "";
    this.monthsAtBank = 0;
    this.nameOnBankAccount = "";
    this.accountNumber = "";
    this.sortCode = "";
    this.netMonthlyIncome = 0;
  }
}
