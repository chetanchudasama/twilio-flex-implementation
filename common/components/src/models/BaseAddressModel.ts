export class BaseAddressModel {
  subBuilding: string;

  buildingNumber: string;

  buildingName: string;

  streetName: string;

  town: string;

  postcode: string;

  constructor() {
    this.subBuilding = "";
    this.streetName = "";
    this.buildingName = "";
    this.buildingNumber = "";
    this.town = "";
    this.postcode = "";
  }
}
