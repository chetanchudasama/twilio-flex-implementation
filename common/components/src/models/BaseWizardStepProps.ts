export interface BaseWizardStepProps {
  moveForward?: () => void;
  moveBackward?: () => void;
  nextStep?: string;
  previousStep?: string;
}
