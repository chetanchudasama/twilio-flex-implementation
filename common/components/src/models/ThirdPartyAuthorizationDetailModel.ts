import { ThirdPartyRemovalAuthor } from "../shared/enum";

export class ThirdPartyAuthorizationDetail {
  public thirdPartyName: string;

  public thirdPartyDateOfBirth: Date | null;

  public thirdPartyAddress: string;

  public thirdPartyPostcode: string;

  public thirdPartyExpiryDate: Date | null;

  public thirdPartyDoNotDiscuss: string;

  public thirdPartyRemovalAuthor: ThirdPartyRemovalAuthor | string;

  public thirdPartyRemovalNotes: string;

  constructor() {
    this.thirdPartyName = "";
    this.thirdPartyDateOfBirth = null;
    this.thirdPartyAddress = "";
    this.thirdPartyPostcode = "";
    this.thirdPartyExpiryDate = null;
    this.thirdPartyDoNotDiscuss = "";
    this.thirdPartyRemovalAuthor = "";
    this.thirdPartyRemovalNotes = "";
  }
}
