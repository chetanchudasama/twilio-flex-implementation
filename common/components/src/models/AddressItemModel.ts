/* eslint-disable max-classes-per-file */
import { BaseAddressModel } from "./BaseAddressModel";

class AddressStatusItemModel {
  addressStatusId: number;

  addressStatusName: string;

  yearsAtAddress: number;

  monthsAtAddress: number;

  constructor() {
    this.addressStatusId = -1;
    this.addressStatusName = "";
    this.yearsAtAddress = 0;
    this.monthsAtAddress = 0;
  }
}

export class AddressItemModel extends BaseAddressModel {
  addressStatus: AddressStatusItemModel;

  isPrimaryAddress: boolean;

  public constructor() {
    super();
    this.addressStatus = new AddressStatusItemModel();
    this.isPrimaryAddress = false;
  }
}
