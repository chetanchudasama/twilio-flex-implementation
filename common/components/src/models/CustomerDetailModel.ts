import { Shared } from "../shared/Shared";
import { AddressItemModel } from "./AddressItemModel";
import { ApplicationStatusModel } from "./ApplicationStatusModel";
import { ThirdPartyAuthorizationDetail } from "./ThirdPartyAuthorizationDetailModel";
import { VulnerableCustomerInformation } from "./VulnerableCustomerInformation";
import { PreferredLenderModel } from "./PreferredLenderModel";
import { BankDetailModel } from "./BankDetailModel";
import { CustomerEmploymentModel } from "./CustomerEmploymentModel";
import { TitleModel } from "./TitleModel";
import { DrivingLicenceTypeModel } from "./DrivingLicenceTypeModel";
import { MaritalStatusModel } from "./MaritalStatusModel";

export class CustomerDetailModel {
  applicationId: number;

  title: TitleModel;

  firstName: string;

  middleName: string;

  lastName: string;

  dateOfBirth!: Date;

  applicationDate!: Date;

  amountToFinance: number;

  monthlyBudget: number;

  savedCarsCount: number;

  callbackBooked!: Date;

  mobileNumber: string;

  workNumber: string;

  homeNumber: string;

  emailAddress: string;

  maritalStatus: MaritalStatusModel;

  drivingLicenceType: DrivingLicenceTypeModel;

  pin: string;

  applicationStatus: ApplicationStatusModel;

  thirdPartyAuthorization: ThirdPartyAuthorizationDetail | null;

  hasThirdPartyAuthorization: boolean;

  hasVulnerableCustomerReported: boolean;

  vulnerableCustomerInformation: VulnerableCustomerInformation | null;

  currentProgressStatus: string;

  preferredLender: PreferredLenderModel;

  bankDetails: BankDetailModel;

  addresses: AddressItemModel[];

  employmentHistory: CustomerEmploymentModel[];

  constructor() {
    this.applicationId = -1;
    this.title = new TitleModel();
    this.firstName = "";
    this.middleName = "";
    this.lastName = "";
    this.amountToFinance = 0;
    this.monthlyBudget = 0;
    this.savedCarsCount = 0;
    this.mobileNumber = "";
    this.workNumber = "";
    this.homeNumber = "";
    this.emailAddress = "";
    this.maritalStatus = new MaritalStatusModel();
    this.drivingLicenceType = new DrivingLicenceTypeModel();
    this.pin = "";
    this.applicationStatus = new ApplicationStatusModel();
    this.thirdPartyAuthorization = null;
    this.hasThirdPartyAuthorization = false;
    this.hasVulnerableCustomerReported = false;
    this.vulnerableCustomerInformation = null;
    this.currentProgressStatus = "";
    this.preferredLender = new PreferredLenderModel();
    this.bankDetails = new BankDetailModel();
    this.addresses = [];
    this.employmentHistory = [];
  }

  get fullName(): string {
    const { title, firstName, middleName, lastName } = this;
    const currentTitle = title && title.titleName || "";

    return Shared.getFormattedText(
      [currentTitle, firstName, middleName, lastName],
      " "
    );
  }

  get primaryAddress(): AddressItemModel {
    const { addresses } = this;
    return (
      addresses.find((address: AddressItemModel) => address.isPrimaryAddress) ||
      new AddressItemModel()
    );
  }

  get address(): string {
    const { primaryAddress } = this;

    if (primaryAddress) {
      const addressList = [
        primaryAddress.buildingNumber
          ? primaryAddress.buildingNumber.toString()
          : "",
        primaryAddress.buildingName,
        primaryAddress.subBuilding,
        primaryAddress.streetName,
        primaryAddress.town,
        primaryAddress.postcode,
      ];
      return Shared.getFormattedText(addressList, ", ");
    }
    return "";
  }
}
