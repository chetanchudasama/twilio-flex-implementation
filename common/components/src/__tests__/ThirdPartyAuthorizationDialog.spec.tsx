/* eslint-disable react/jsx-props-no-spreading */
import { create, act } from "react-test-renderer";
import { shallow } from "enzyme";
import React from "react";
import {
  ThirdPartyAuthorizationDialog,
  ThirdPartyAuthorizationDialogProps,
} from "../ThirdPartyAuthorizationDialog/ThirdPartyAuthorizationDialog";
import { ThirdPartyAuthorizationDetail } from "../models/ThirdPartyAuthorizationDetailModel";
import { ResponsiveDialog } from "../ResponsiveDialog";

const props: ThirdPartyAuthorizationDialogProps = {
  open: true,
  thirdPartyAuthorizationInformation: new ThirdPartyAuthorizationDetail(),
  handleDialogClose: jest.fn(() => {}),
  handleSaveThirdPartyDetail: jest.fn(() => {}),
};

jest.mock("../ResponsiveDialog/index", () => {
  return {
    ResponsiveDialog: jest.fn(({ children }) => children),
  };
});

jest.mock("../MuiDatePicker/MuiDatePicker", () => {
  return {
    MuiDatePicker: () => {
      return <div>Date Picker</div>;
    },
  };
});

jest.mock("@material-ui/core", () => {
  const materialUI = jest.requireActual("@material-ui/core");
  return {
    ...materialUI,
    FormControlLabel: () => {
      return <div>Form Control</div>;
    },
  };
});

describe("ThirdPartyAuthorizationDialog", () => {
  it("renders correctly", () => {
    const wrapper = create(
      <ThirdPartyAuthorizationDialog {...props} />
    ).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it("should call handleSaveThirdPartyDetail method, when Save button is clicked", () => {
    props.thirdPartyAuthorizationInformation = {
      thirdPartyName: "Name",
      thirdPartyDateOfBirth: new Date(),
      thirdPartyAddress: "Test",
      thirdPartyPostcode: "Test",
      thirdPartyExpiryDate: new Date(),
      thirdPartyDoNotDiscuss: "Test",
      thirdPartyRemovalAuthor: "",
      thirdPartyRemovalNotes: "Test",
    };
    const wrapper = shallow(<ThirdPartyAuthorizationDialog {...props} />);

    const spy = jest.spyOn(props, "handleSaveThirdPartyDetail");

    act(() => {
      wrapper.children().props().onConfirm();
    });

    expect(spy).toHaveBeenCalled();
  });

  it("should call handleDialogClose method, when cancel button is clicked", () => {
    const wrapper = shallow(<ThirdPartyAuthorizationDialog {...props} />);

    const spy = jest.spyOn(props, "handleDialogClose");

    act(() => {
      wrapper.children().props().onCancel();
    });

    expect(spy).toHaveBeenCalled();
  });

  it("should render fields as disabled, when third party detail is saved", () => {
    props.thirdPartyAuthorizationInformation = {
      thirdPartyName: "Name",
      thirdPartyDateOfBirth: new Date(),
      thirdPartyAddress: "Test",
      thirdPartyPostcode: "Test",
      thirdPartyExpiryDate: new Date(),
      thirdPartyDoNotDiscuss: "Test",
      thirdPartyRemovalAuthor: "",
      thirdPartyRemovalNotes: "Test",
    };
    const wrapper = create(<ThirdPartyAuthorizationDialog {...props} />).root;

    expect(
      wrapper.findByProps({
        className: "name",
      }).props.disabled
    ).toEqual(true);
    expect(
      wrapper.findByProps({
        className: "address",
      }).props.disabled
    ).toEqual(true);
    expect(
      wrapper.findByProps({
        className: "postcode",
      }).props.disabled
    ).toEqual(true);
    expect(
      wrapper.findByProps({
        className: "do-not-discuss",
      }).props.disabled
    ).toEqual(true);
  });

  it("should render confirmation dialog, when delete button is clicked", () => {
    const wrapper = shallow(<ThirdPartyAuthorizationDialog {...props} />);

    expect(wrapper.find(ResponsiveDialog)).toHaveLength(1);
    act(() => {
      wrapper.children().props().onDelete();
    });

    expect(wrapper.find(ResponsiveDialog)).toHaveLength(2);
    expect(wrapper.find(".confirmation-message").exists()).toBeTruthy();
    expect(wrapper.find(".confirmation-message").text()).toContain(
      "Are you sure you want to clear the third party authorization details?"
    );
  });

  it("should close confirmation dialog, when No button is clicked from confirmation dialog", () => {
    const wrapper = shallow(<ThirdPartyAuthorizationDialog {...props} />);

    act(() => {
      wrapper.children().props().onDelete();
    });

    expect(wrapper.find(".confirmation-message").exists()).toBeTruthy();

    act(() => {
      wrapper.find(ResponsiveDialog).at(1).props().onCancel();
    });

    expect(wrapper.find(".confirmation-message").exists()).toBeFalsy();
  });

  it("should call handleSaveThirdPartyDetail method, when Yes button is clicked from confirmation dialog", () => {
    const wrapper = shallow(<ThirdPartyAuthorizationDialog {...props} />);

    const spy = jest.spyOn(props, "handleSaveThirdPartyDetail");

    act(() => {
      wrapper.children().props().onDelete();
    });

    expect(wrapper.find(".confirmation-message").exists()).toBeTruthy();

    act(() => {
      wrapper.find(ResponsiveDialog).at(1).props().onConfirm();
    });

    expect(spy).toHaveBeenCalledWith(null);
  });
});
