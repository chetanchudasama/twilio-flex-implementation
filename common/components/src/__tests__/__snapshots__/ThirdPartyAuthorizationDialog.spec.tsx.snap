// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThirdPartyAuthorizationDialog renders correctly 1`] = `
<div>
  <div
    className="css-mof3sf"
  >
    <div
      className="MuiGrid-container-1 third-party-permission-container"
    >
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41"
      >
        <p>
          You are leaving third-party permission for this customer. All fields are required
        </p>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 form-input"
      >
        <div
          className="MuiFormControl-root-98 MuiFormControl-fullWidth-101 name"
        >
          <label
            className="MuiFormLabel-root-113 MuiFormLabel-required-118 MuiInputLabel-required-106 MuiInputLabel-root-102 MuiInputLabel-formControl-107 MuiInputLabel-animated-110"
            data-shrink={false}
          >
            Name
            <span
              className="MuiFormLabel-asterisk-119"
            >
               *
            </span>
          </label>
          <div
            className="MuiInputBase-root-133 MuiInput-root-120 MuiInput-underline-124 MuiInputBase-fullWidth-142 MuiInput-fullWidth-127 MuiInputBase-formControl-134 MuiInput-formControl-121"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-143 MuiInput-input-128"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="text"
              value=""
            />
          </div>
          
        </div>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 form-input"
      >
        <div
          className="MuiFormControl-root-98 MuiFormControl-fullWidth-101 address"
        >
          <label
            className="MuiFormLabel-root-113 MuiFormLabel-required-118 MuiInputLabel-required-106 MuiInputLabel-root-102 MuiInputLabel-formControl-107 MuiInputLabel-animated-110"
            data-shrink={false}
          >
            Address
            <span
              className="MuiFormLabel-asterisk-119"
            >
               *
            </span>
          </label>
          <div
            className="MuiInputBase-root-133 MuiInput-root-120 MuiInput-underline-124 MuiInputBase-fullWidth-142 MuiInput-fullWidth-127 MuiInputBase-formControl-134 MuiInput-formControl-121 MuiInputBase-multiline-141 MuiInput-multiline-126"
            onClick={[Function]}
          >
            <textarea
              aria-invalid={false}
              className="MuiInputBase-input-143 MuiInput-input-128 MuiInputBase-inputMultiline-145 MuiInput-inputMultiline-130"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              rows={2}
              value=""
            />
          </div>
          
        </div>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 form-input"
      >
        <div
          className="MuiFormControl-root-98 MuiFormControl-fullWidth-101 postcode"
        >
          <label
            className="MuiFormLabel-root-113 MuiFormLabel-required-118 MuiInputLabel-required-106 MuiInputLabel-root-102 MuiInputLabel-formControl-107 MuiInputLabel-animated-110"
            data-shrink={false}
          >
            Postcode
            <span
              className="MuiFormLabel-asterisk-119"
            >
               *
            </span>
          </label>
          <div
            className="MuiInputBase-root-133 MuiInput-root-120 MuiInput-underline-124 MuiInputBase-fullWidth-142 MuiInput-fullWidth-127 MuiInputBase-formControl-134 MuiInput-formControl-121"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-143 MuiInput-input-128"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="text"
              value=""
            />
          </div>
          
        </div>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-6-35 form-input padding-right"
      >
        <div>
          Date Picker
        </div>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-6-35 form-input padding-left"
      >
        <div>
          Date Picker
        </div>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 form-input"
      >
        <div
          className="MuiFormControl-root-98 MuiFormControl-fullWidth-101 do-not-discuss"
        >
          <label
            className="MuiFormLabel-root-113 MuiFormLabel-required-118 MuiInputLabel-required-106 MuiInputLabel-root-102 MuiInputLabel-formControl-107 MuiInputLabel-animated-110"
            data-shrink={false}
          >
            Do Not discuss
            <span
              className="MuiFormLabel-asterisk-119"
            >
               *
            </span>
          </label>
          <div
            className="MuiInputBase-root-133 MuiInput-root-120 MuiInput-underline-124 MuiInputBase-fullWidth-142 MuiInput-fullWidth-127 MuiInputBase-formControl-134 MuiInput-formControl-121 MuiInputBase-multiline-141 MuiInput-multiline-126"
            onClick={[Function]}
          >
            <textarea
              aria-invalid={false}
              className="MuiInputBase-input-143 MuiInput-input-128 MuiInputBase-inputMultiline-145 MuiInput-inputMultiline-130"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              rows={2}
              value=""
            />
          </div>
          
        </div>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 form-input"
      >
        <label
          className="MuiFormLabel-root-113 MuiFormLabel-required-118 MuiInputLabel-required-106 MuiInputLabel-root-102 MuiInputLabel-animated-110"
        >
          Third Party Removal Author
          <span
            className="MuiFormLabel-asterisk-119"
          >
             *
          </span>
        </label>
        <div
          className="MuiFormControl-root-98 MuiFormControl-fullWidth-101"
        >
          <div
            className="MuiFormGroup-root-150 radio-group"
            role="radiogroup"
          >
            <div>
              Form Control
            </div>
            <div>
              Form Control
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
